File: Core/Stateflows/Common/Classes/Watcher.cs
using System;
using System.Threading.Tasks;

namespace Stateflows.Common.Classes
{
    public sealed class Watcher : IWatcher
    {
        private readonly IUnwatcher unwatcher;

        public Watcher(IUnwatcher unwatcher)
        {
            this.unwatcher = unwatcher;
        }

        public async ValueTask DisposeAsync()
            => await unwatcher.UnwatchAsync(this);
        
        public void Dispose()
        {
            Dispose(true);
            GC.SuppressFinalize(this);
        }

        private void Dispose(bool disposing)
        {
            unwatcher.UnwatchAsync(this).GetAwaiter().GetResult();
        }

        ~Watcher()
            => Dispose(false);
    }
}

File: Transport/@stateflows/common/src/classes/watcher.ts
import { IWatcher } from "../interfaces/watcher";
import { IUnwatcher } from "../interfaces/unwatcher";

export class Watcher implements IWatcher, Disposable {
    constructor(
        private notificationName: string,
        private unwatcher: IUnwatcher
    ) {}

    [Symbol.dispose](): void {
        this.unwatcher.unwatch(this.notificationName);
    }

    async unwatch(): Promise<void> {
        this[Symbol.dispose]();
    }
}
